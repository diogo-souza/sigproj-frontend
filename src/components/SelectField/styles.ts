import styled, { css, DefaultTheme } from 'styled-components';

import { SelectFieldProps } from '.';

type ContainerProps = Pick<SelectFieldProps, 'disabled'> & {
  error?: boolean;
};

export const Select = styled.select`
  ${({ theme }) => css`
    color: ${theme.colors.titleBlack};
    cursor: pointer;
    font-family: ${theme.font.family};
    font-size: ${theme.font.sizes.medium};
    font-weight: ${theme.font.semiBold};
    appearance: none;
    border: 0;
    outline: none;
    width: 100%;
    padding: ${theme.spacings.xxsmall} ${theme.spacings.large}
      ${theme.spacings.xxsmall} 12px;
    background: transparent;
  `}
`;

export const SelectContainer = styled.div`
  ${({ theme }) => css`
    position: relative;

    border: 0.1rem solid;
    background: ${theme.colors.white};
    border-radius: ${theme.border.radius};

    border-color: ${theme.colors.gray};
    &:hover {
      border-color: ${theme.colors.darkGray};
    }
    > svg {
      color: ${theme.colors.titleBlack};

      position: absolute;
      right: 12px;
      top: 50%;
      transform: translateY(-50%);

      width: 2rem;
      height: 1.2rem;
      pointer-events: none;
    }
    &:focus-within {
      border-color: ${theme.colors.focus};
      box-shadow: 0 0 0 0.3rem ${theme.colors.focusShadow};
    }
  `}
`;

export const Error = styled.p`
  ${({ theme }) => css`
    color: ${theme.colors.danger};
    font-size: ${theme.font.sizes.small};
  `};
`;

const containerModifiers = {
  disabled: (theme: DefaultTheme) => css`
    ${SelectContainer} {
      background: ${theme.colors.lightGray};
    }
    ${Select},
    ${SelectContainer},
    option {
      cursor: not-allowed;
      color: ${theme.colors.darkGray};
      &::placeholder {
        color: currentColor;
      }
    }
  `,
  error: (theme: DefaultTheme) => css`
    ${SelectContainer} {
      border-color: ${theme.colors.danger};
      &:focus-within {
        box-shadow: 0 0 0 0.3rem ${theme.colors.dangerShadow};
      }
    }
    svg {
      color: ${theme.colors.danger};
    }
  `,
};

export const Container = styled.div<ContainerProps>`
  ${({ theme, disabled, error }) => css`
    ${error && containerModifiers.error(theme)}
    ${disabled && containerModifiers.disabled(theme)}
  `}
`;
