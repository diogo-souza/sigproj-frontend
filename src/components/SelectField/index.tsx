import React, { SelectHTMLAttributes, useState } from 'react';
import { FaChevronDown as ArrowDownIcon } from 'react-icons/fa';

import Label from 'components/Label';
import * as S from './styles';

export type SelectFieldProps = {
  label?: string;
  initialValue?: string;
  disabled?: boolean;
  error?: string;
  options?: {
    value: string;
    text: string;
  }[];
} & SelectHTMLAttributes<HTMLSelectElement>;

const SelectField: React.FC<SelectFieldProps> = ({
  name,
  label,
  placeholder,
  initialValue = '',
  disabled = false,
  error,
  style,
  className,
  options,
  ...props
}: SelectFieldProps) => {
  const [selected, setSelected] = useState(initialValue);

  const onChange = (e: React.ChangeEvent<HTMLSelectElement>) => {
    const newValue = e.currentTarget.value;
    setSelected(newValue);
  };

  return (
    <S.Container
      disabled={disabled}
      error={!!error}
      style={style}
      className={className}
    >
      {!!label && (
        <Label htmlFor={name} disabled={disabled} hasError={!!error}>
          {label}
        </Label>
      )}
      <S.SelectContainer>
        <S.Select
          name={name}
          value={selected}
          onChange={onChange}
          disabled={disabled}
          {...(label ? { id: name } : {})}
          {...props}
        >
          {!!placeholder && <option value="">{placeholder}</option>}
          {options?.map(({ value, text }) => (
            <option key={value} value={value}>
              {text}
            </option>
          ))}
        </S.Select>
        <ArrowDownIcon />
      </S.SelectContainer>
      {!!error && <S.Error role="alert">{error}</S.Error>}
    </S.Container>
  );
};

export default SelectField;
