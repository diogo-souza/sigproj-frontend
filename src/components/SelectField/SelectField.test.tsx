import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { renderWithTheme } from 'utils/tests/helpers';

import SelectField from '.';

describe('<SelectField />', () => {
  it('should render the heading', () => {
    renderWithTheme(<SelectField label="Label" name="Field" />);
    expect(screen.getByLabelText('Label')).toBeInTheDocument();
  });

  it('should render without Label', () => {
    renderWithTheme(<SelectField />);
    expect(screen.queryByLabelText('Label')).not.toBeInTheDocument();
  });

  it('should render with placeholder', () => {
    renderWithTheme(<SelectField placeholder="Hey you" />);

    expect(screen.getByRole('option', { name: /hey you/i })).toHaveValue('');
  });

  it('should changes its value when select', async () => {
    renderWithTheme(
      <SelectField
        data-testid="select"
        label="TextField"
        name="textfield"
        options={[
          { value: 'test 1', text: 'Teste 1' },
          { value: 'test 2', text: 'Teste 2' },
          { value: 'test 3', text: 'Teste 3' },
        ]}
      />,
    );

    const select = screen.getByTestId('select') as HTMLSelectElement;
    userEvent.selectOptions(select, 'test 3');
    const options = screen.getAllByRole('option') as HTMLOptionElement[];

    expect(options[0].selected).toBeFalsy();
    expect(options[1].selected).toBeFalsy();
    expect(options[2].selected).toBeTruthy();
  });

  it('should not changes its value when disabled', async () => {
    renderWithTheme(
      <SelectField
        data-testid="select"
        label="TextField"
        name="textfield"
        disabled
        options={[
          { value: 'test 1', text: 'Teste 1' },
          { value: 'test 2', text: 'Teste 2' },
          { value: 'test 3', text: 'Teste 3' },
        ]}
      />,
    );

    const select = screen.getByTestId('select') as HTMLSelectElement;
    expect(select).toBeDisabled();

    userEvent.selectOptions(select, 'test 3');
    const options = screen.getAllByRole('option') as HTMLOptionElement[];

    expect(options[0].selected).toBeTruthy();
    expect(options[1].selected).toBeFalsy();
    expect(options[2].selected).toBeFalsy();
  });

  it('should render with error', () => {
    const { container } = renderWithTheme(
      <SelectField label="TextField" name="TextField" error="Error message" />,
    );

    expect(screen.getByText('Error message')).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should be accessible by tab', () => {
    renderWithTheme(
      <SelectField data-testid="select" label="TextField" name="TextField" />,
    );

    const select = screen.getByTestId('select') as HTMLSelectElement;
    expect(document.body).toHaveFocus();

    userEvent.tab();
    expect(select).toHaveFocus();
  });

  it('should not accessible by tab when disabled', () => {
    renderWithTheme(
      <SelectField
        data-testid="select"
        label="TextField"
        name="TextField"
        disabled
      />,
    );

    const select = screen.getByTestId('select') as HTMLSelectElement;
    expect(document.body).toHaveFocus();

    userEvent.tab();
    expect(select).not.toHaveFocus();
  });
});
