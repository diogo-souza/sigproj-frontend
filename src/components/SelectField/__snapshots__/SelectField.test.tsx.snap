// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SelectField /> should render with error 1`] = `
.c1 {
  display: inline-block;
  font-size: 1.6rem;
  color: #171923;
  font-weight: 600;
  cursor: pointer;
  margin-bottom: 0.8rem;
  color: #dc3545;
}

.c4 {
  color: #171923;
  cursor: pointer;
  font-family: Open-Sans,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Helvetica Neue',sans-serif;
  font-size: 1.6rem;
  font-weight: 500;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border: 0;
  outline: none;
  width: 100%;
  padding: 0.8rem 4.0rem 0.8rem 12px;
  background: transparent;
}

.c3 {
  position: relative;
  border: 0.1rem solid;
  background: #FFFFFF;
  border-radius: 0.4rem;
  border-color: #939393;
}

.c3:hover {
  border-color: #212227;
}

.c3 > svg {
  color: #171923;
  position: absolute;
  right: 12px;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  width: 2rem;
  height: 1.2rem;
  pointer-events: none;
}

.c3:focus-within {
  border-color: #0d6efd;
  box-shadow: 0 0 0 0.3rem rgba(75,174,227,0.4);
}

.c5 {
  color: #dc3545;
  font-size: 1.4rem;
}

.c0 .c2 {
  border-color: #dc3545;
}

.c0 .c2:focus-within {
  box-shadow: 0 0 0 0.3rem rgba(220,53,69,0.4);
}

.c0 svg {
  color: #dc3545;
}

<div
  class="c0"
>
  <label
    class="c1"
    for="TextField"
  >
    TextField
  </label>
  <div
    class="c2 c3"
  >
    <select
      class="c4"
      id="TextField"
      name="TextField"
    />
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 448 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
      />
    </svg>
  </div>
  <p
    class="c5"
    role="alert"
  >
    Error message
  </p>
</div>
`;
